File: ios/Plugin/CallManager.swift
import Foundation
import WebRTC
import OneSignal
import ReplayKit

enum CallError: Error {
    case invalidState
    case connectionFailed
    case recordingFailed
    case invalidConfiguration
    case securityError
}

class CallManager: NSObject {
    private weak var plugin: VoiceVideoCallPlugin?
    private var peerConnectionFactory: RTCPeerConnectionFactory?
    private var peerConnection: RTCPeerConnection?
    private var mediaStream: RTCMediaStream?
    private var localVideoTrack: RTCVideoTrack?
    private var localAudioTrack: RTCAudioTrack?
    private var screenCapturer: RTCVideoCapturer?
    private var recorder: RPScreenRecorder?
    private var isScreenSharing = false
    private var isRecording = false
    private var isAudioMuted = false
    private var isVideoMuted = false
    private var participants: [String: Participant] = [:]
    private var metricsCollector: MetricsCollector?
    
    init(plugin: VoiceVideoCallPlugin) {
        super.init()
        self.plugin = plugin
        self.setupWebRTC()
        self.metricsCollector = MetricsCollector()
    }
    
    private func setupWebRTC() {
        RTCInitializeSSL()
        let videoEncoderFactory = RTCDefaultVideoEncoderFactory()
        let videoDecoderFactory = RTCDefaultVideoDecoderFactory()
        peerConnectionFactory = RTCPeerConnectionFactory(
            encoderFactory: videoEncoderFactory,
            decoderFactory: videoDecoderFactory
        )
    }
    
    // MARK: - Call Management
    func startCall(roomId: String, callType: String, completion: @escaping (Error?) -> Void) {
        // Implementation for starting a call
        completion(nil)
    }
    
    func acceptCall(roomId: String, completion: @escaping (Error?) -> Void) {
        // Implementation for accepting a call
        completion(nil)
    }
    
    func rejectCall(roomId: String, completion: @escaping (Error?) -> Void) {
        // Implementation for rejecting a call
        completion(nil)
    }
    
    func endCall(completion: @escaping (Error?) -> Void) {
        // Implementation for ending a call
        completion(nil)
    }
    
    // MARK: - Screen Sharing
    func startScreenShare(withAudio: Bool, completion: @escaping (Error?) -> Void) {
        let recorder = RPScreenRecorder.shared()
        guard recorder.isAvailable else {
            completion(CallError.invalidState)
            return
        }
        
        recorder.startCapture { [weak self] buffer, type, error in
            guard let self = self else { return }
            if let error = error {
                completion(error)
                return
            }
            
            // Convert buffer to RTCVideoFrame and send via WebRTC
            self.isScreenSharing = true
            self.plugin?.notifyScreenShareStateChanged(isSharing: true)
        }
    }
    
    func stopScreenShare(completion: @escaping (Error?) -> Void) {
        RPScreenRecorder.shared().stopCapture { [weak self] error in
            guard let self = self else { return }
            if let error = error {
                completion(error)
                return
            }
            
            self.isScreenSharing = false
            self.plugin?.notifyScreenShareStateChanged(isSharing: false)
            completion(nil)
        }
    }
    
    // MARK: - Recording
    func startRecording(type: String, quality: String, storage: String, completion: @escaping (Error?) -> Void) {
        guard !isRecording else {
            completion(CallError.invalidState)
            return
        }
        
        // Configure and start recording based on parameters
        isRecording = true
        plugin?.notifyRecordingStateChanged("started")
        completion(nil)
    }
    
    func stopRecording(completion: @escaping (Result<String, Error>) -> Void) {
        guard isRecording else {
            completion(.failure(CallError.invalidState))
            return
        }
        
        // Stop recording and return file path
        isRecording = false
        plugin?.notifyRecordingStateChanged("stopped")
        completion(.success("recording_path"))
    }
    
    // MARK: - Media Controls
    func muteAudio(completion: @escaping (Error?) -> Void) {
        localAudioTrack?.isEnabled = false
        isAudioMuted = true
        completion(nil)
    }
    
    func unmuteAudio(completion: @escaping (Error?) -> Void) {
        localAudioTrack?.isEnabled = true
        isAudioMuted = false
        completion(nil)
    }
    
    func muteVideo(completion: @escaping (Error?) -> Void) {
        localVideoTrack?.isEnabled = false
        isVideoMuted = true
        completion(nil)
    }
    
    func unmuteVideo(completion: @escaping (Error?) -> Void) {
        localVideoTrack?.isEnabled = true
        isVideoMuted = false
        completion(nil)
    }
    
    func switchCamera(completion: @escaping (Error?) -> Void) {
        // Implementation for camera switching
        completion(nil)
    }
    
    // MARK: - Multi-Party Call
    func addParticipant(participantId: String, completion: @escaping (Error?) -> Void) {
        let participant = Participant(id: participantId)
        participants[participantId] = participant
        plugin?.notifyParticipantStateChanged(participant.toDictionary(), joined: true)
        completion(nil)
    }
    
    func removeParticipant(participantId: String, completion: @escaping (Error?) -> Void) {
        participants.removeValue(forKey: participantId)
        plugin?.notifyParticipantStateChanged(["id": participantId], joined: false)
        completion(nil)
    }
    
    func getParticipants(completion: @escaping (Result<[[String: Any]], Error>) -> Void) {
        let participantsList = participants.values.map { $0.toDictionary() }
        completion(.success(participantsList))
    }
    
    // MARK: - Network
    func setNetworkConfig(config: [String: Any], completion: @escaping (Error?) -> Void) {
        // Implementation for network configuration
        completion(nil)
    }
    
    func getNetworkStatus(completion: @escaping (Result<[String: Any], Error>) -> Void) {
        // Implementation for network status
        completion(.success([:]))
    }
    
    // MARK: - Security
    func setSecurityOptions(options: [String: Any], completion: @escaping (Error?) -> Void) {
        // Implementation for security options
        completion(nil)
    }
    
    func generateCallToken(completion: @escaping (Result<String, Error>) -> Void) {
        // Implementation for token generation
        completion(.success("generated_token"))
    }
    
    func validateCallToken(token: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        // Implementation for token validation
        completion(.success(true))
    }
    
    // MARK: - Analytics
    func getCallMetrics(completion: @escaping (Result<[String: Any], Error>) -> Void) {
        completion(.success(metricsCollector?.getMetrics() ?? [:]))
    }
    
    func exportCallLogs(completion: @escaping (Result<String, Error>) -> Void) {
        completion(.success(metricsCollector?.exportLogs() ?? ""))
    }
    
    // MARK: - Background Mode
    func setBackgroundMode(options: [String: Any], completion: @escaping (Error?) -> Void) {
        // Implementation for background mode
        completion(nil)
    }
}

// MARK: - Helper Classes
class Participant {
    let id: String
    var connection: RTCPeerConnection?
    var stream: RTCMediaStream?
    
    init(id: String) {
        self.id = id
    }
    
    func toDictionary() -> [String: Any] {
        return [
            "id": id,
            "hasAudio": stream?.audioTracks.count ?? 0 > 0,
            "hasVideo": stream?.videoTracks.count ?? 0 > 0
        ]
    }
}

class MetricsCollector {
    private var metrics: [[String: Any]] = []
    
    func addMetric(_ metric: [String: Any]) {
        metrics.append(metric)
    }
    
    func getMetrics() -> [String: Any] {
        return [
            "metrics": metrics,
            "timestamp": Date().timeIntervalSince1970
        ]
    }
    
    func exportLogs() -> String {
        return JSONSerialization.stringify(metrics) ?? ""
    }
}

// MARK: - Extensions
extension JSONSerialization {
    static func stringify(_ value: Any) -> String? {
        if let data = try? JSONSerialization.data(withJSONObject: value, options: [.prettyPrinted]) {
            return String(data: data, encoding: .utf8)
        }
        return nil
    }
}

File: android/src/main/java/com/example/voicevideocall/CallManager.java
package com.example.voicevideocall;

import android.content.Context;
import android.media.MediaRecorder;
import android.util.Log;
import org.webrtc.*;
import com.onesignal.OneSignal;
import org.json.JSONObject;
import java.util.*;

public class CallManager {
    private static final String TAG = "CallManager";
    
    private Context context;
    private PeerConnectionFactory peerConnectionFactory;
    private PeerConnection peerConnection;
    private MediaStream mediaStream;
    private VideoTrack localVideoTrack;
    private AudioTrack localAudioTrack;
    private MediaRecorder mediaRecorder;
    private boolean isScreenSharing = false;
    private boolean isRecording = false;
    private boolean isAudioMuted = false;
    private boolean isVideoMuted = false;
    private List<Participant> participants = new ArrayList<>();
    private CallMetricsCollector metricsCollector;

    // Callback interfaces
    public interface CallEventListener {
        void onCallQualityChanged(JSONObject stats);
        void onScreenShareStateChanged(boolean isSharing);
        void onRecordingStateChanged(String state);
        void onParticipantStateChanged(JSONObject participant, boolean joined);
        void onNetworkStatusChanged(JSONObject status);
        void onMetricsUpdated(JSONObject metrics);
    }

    private CallEventListener eventListener;

    public CallManager(Context context, CallEventListener listener) {
        this.context = context;
        this.eventListener = listener;
        initializePeerConnectionFactory();
        metricsCollector = new CallMetricsCollector();
    }

    private void initializePeerConnectionFactory() {
        PeerConnectionFactory.InitializationOptions initializationOptions =
                PeerConnectionFactory.InitializationOptions.builder(context)
                        .setEnableInternalTracer(true)
                        .createInitializationOptions();
        PeerConnectionFactory.initialize(initializationOptions);

        PeerConnectionFactory.Options options = new PeerConnectionFactory.Options();
        peerConnectionFactory = PeerConnectionFactory.builder()
                .setOptions(options)
                .createPeerConnectionFactory();
    }

    // Screen Sharing Methods
    public void startScreenShare(boolean withAudio) {
        if (!isScreenSharing) {
            // Implementation for screen capture
            MediaProjectionManager projectionManager = 
                (MediaProjectionManager) context.getSystemService(Context.MEDIA_PROJECTION_SERVICE);
            // Request screen capture permission and start sharing
            isScreenSharing = true;
            eventListener.onScreenShareStateChanged(true);
        }
    }

    public void stopScreenShare() {
        if (isScreenSharing) {
            // Stop screen capture
            isScreenSharing = false;
            eventListener.onScreenShareStateChanged(false);
        }
    }

    // Recording Methods
    public void startRecording(String type, String quality, String storage) {
        if (!isRecording) {
            mediaRecorder = new MediaRecorder();
            // Configure mediaRecorder based on type and quality
            mediaRecorder.start();
            isRecording = true;
            eventListener.onRecordingStateChanged("started");
        }
    }

    public String stopRecording() {
        if (isRecording) {
            mediaRecorder.stop();
            mediaRecorder.release();
            mediaRecorder = null;
            isRecording = false;
            eventListener.onRecordingStateChanged("stopped");
            return "recording_path"; // Return actual recording path
        }
        return null;
    }

    // Media Control Methods
    public void muteAudio() {
        if (localAudioTrack != null) {
            localAudioTrack.setEnabled(false);
            isAudioMuted = true;
        }
    }

    public void unmuteAudio() {
        if (localAudioTrack != null) {
            localAudioTrack.setEnabled(true);
            isAudioMuted = false;
        }
    }

    public void muteVideo() {
        if (localVideoTrack != null) {
            localVideoTrack.setEnabled(false);
            isVideoMuted = true;
        }
    }

    public void unmuteVideo() {
        if (localVideoTrack != null) {
            localVideoTrack.setEnabled(true);
            isVideoMuted = false;
        }
    }

    public void switchCamera() {
        // Implementation for camera switching
    }

    // Multi-Party Call Methods
    public void addParticipant(String participantId) {
        Participant participant = new Participant(participantId);
        participants.add(participant);
        // Create new peer connection for this participant
        JSONObject participantData = participant.toJSON();
        eventListener.onParticipantStateChanged(participantData, true);
    }

    public void removeParticipant(String participantId) {
        participants.removeIf(p -> p.getId().equals(participantId));
        eventListener.onParticipantStateChanged(
            new JSONObject().put("id", participantId), 
            false
        );
    }

    public List<Participant> getParticipants() {
        return new ArrayList<>(participants);
    }

    // Network Methods
    public void setNetworkConfig(JSONObject config) {
        // Update ICE servers and connection config
        List<PeerConnection.IceServer> iceServers = parseIceServers(config);
        // Recreate peer connection with new config if needed
    }

    public JSONObject getNetworkStatus() {
        JSONObject status = new JSONObject();
        // Collect current network metrics
        return status;
    }

    // Security Methods
    public void setSecurityOptions(JSONObject options) {
        // Implement E2EE if specified
        // Set up room password
        // Configure allowed domains
    }

    public String generateCallToken() {
        // Generate secure token for call authentication
        return "generated_token";
    }

    public boolean validateCallToken(String token) {
        // Validate provided token
        return true;
    }

    // Analytics Methods
    public JSONObject getCallMetrics() {
        return metricsCollector.getMetrics();
    }

    public String exportCallLogs() {
        return metricsCollector.exportLogs();
    }

    // Background Mode Methods
    public void setBackgroundMode(JSONObject options) {
        // Configure background behavior
        // Set up background notification if needed
    }

    // Helper class for participants
    private static class Participant {
        private String id;
        private PeerConnection connection;
        private MediaStream stream;

        public Participant(String id) {
            this.id = id;
        }

        public String getId() {
            return id;
        }

        public JSONObject toJSON() {
            JSONObject json = new JSONObject();
            // Add participant details
            return json;
        }
    }

    // Helper class for metrics collection
    private static class CallMetricsCollector {
        private List<JSONObject> metrics = new ArrayList<>();

        public void addMetric(JSONObject metric) {
            metrics.add(metric);
        }

        public JSONObject getMetrics() {
            JSONObject allMetrics = new JSONObject();
            // Compile all metrics
            return allMetrics;
        }

        public String exportLogs() {
            // Export metrics as formatted string
            return "";
        }
    }

    private List<PeerConnection.IceServer> parseIceServers(JSONObject config) {
        List<PeerConnection.IceServer> iceServers = new ArrayList<>();
        // Parse ICE server configuration
        return iceServers;
    }
}